### PR TITLE
Add support for arm64-linux (#1)

### PR DESCRIPTION
**Pull request by achichinov**
Add support for arm64-linux

- Copied the latest config.guess
- Copied the latest linux_syscall_support.h from google LSS repository with __aarch64__ support
  Merged some changes from existing coredumper linux_syscall_support.h that are required but absent in the latest LSS.
- Added preprocessor conditions for __aarch64__
- Copied the GP and FP registers structures from glibc.
- Defined FRAME and SET_FRAME macros
  FRAME is defined via getcontext() function. It is simpler than using asm.
- In CreateElfCore() skip [vvar] memory segment.
  A process fails to read this segment. On x86_64 [vvar] is disabled for core dump, but on ARM64 it is not.
  It is the only segment that is marked as "pure PFN range" on the both architectures.
- Make stack pointer 16-byte aligned in local_clone() on all architectures.
- Add a separate ifdef section to read registers using PTRACE_GETREGSET.
  PTRACE_GETREGS and PTRACE_GETFPREGS are not supported on arm64-linux

### CODE REVIEW

File: config.guess [hunk 1]
```diff
@@ -42,7 +42,7 @@ timestamp='2021-06-03'
 # have a pre-POSIX /bin/sh.
 
 
-me=`echo "$0" | sed -e 's,.*/,,'`
+me=$(echo "$0" | sed -e 's,.*/,,')
```
Comment (madscientist) [inline config.guess]: I'm assuming that these changes were made when you used some different version of autoconf, and were not made by hand by you?

File: src/elfcore.c [hunk 2]
```diff
@@ -73,6 +73,8 @@ extern "C" {
     #define O_LARGEFILE 0x2000
   #elif defined(__ARM_ARCH_3__)
     #define O_LARGEFILE 0400000
+  #elif defined(__aarch64__)
+    #define O_LARGEFILE  __O_LARGEFILE
```
Comment (madscientist) [inline src/elfcore.c]: I wonder if it's worthwhile to use something like, **#elif defined (__O_LARGEFILE) \n #define O_LARGEFILE __O_LARGEFILE** just to get all platforms that do define __O_LARGEFILE.

This works too.

File: src/elfcore.c [hunk 3]
```diff
@@ -157,13 +168,15 @@ typedef struct elf_timeval {    /* Time value with microsecond resolution    */
...
+#if !defined(__aarch64__)
 typedef struct elf_siginfo {    /* Information about signal (unused)         */
   int32_t si_signo;             /* Signal number                             */
   int32_t si_code;              /* Extra code                                */
   int32_t si_errno;             /* Errno                                     */
 } elf_siginfo;
-
+#else
+typedef struct elf_siginfo elf_siginfo;
+#endif
```
Comment (madscientist) [inline src/elfcore.c]: Strange.  Why isn't this defined on Intel?

File: src/elfcore.c [hunk 4]
```diff
@@ -1750,6 +1779,28 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
...
+    memset(scratch, 0xFF, sizeof(scratch));
+    struct iovec scratch_iovec = { scratch, sizeof(scratch)};
+    if (sys_ptrace(PTRACE_GETREGSET, pids[i], (void*)NT_PRSTATUS, &scratch_iovec) == 0) {
+      memcpy(thread_regs + i, scratch, sizeof(struct regs));
+      if (main_pid == pids[i]) {
+        SET_FRAME(*(Frame *)frame, thread_regs[i]);
+      }
+      memset(scratch, 0xFF, sizeof(scratch));
+      scratch_iovec.iov_len = sizeof(scratch);
+      if (sys_ptrace(PTRACE_GETREGSET, pids[i], (void*)NT_FPREGSET, &scratch_iovec) == 0) {
```
Comment (madscientist) [inline src/elfcore.c]: ptrace() defines the return code to be -1 on error.  Is it better to check for >=0 here?  Or invert the test and check for <0 and use the goto in the then-block, then you don't need the else statement at all.

File: src/elfcore.c [hunk 5]
```diff
@@ -1785,10 +1836,12 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   /* Get parent's CPU registers, and user data structure                     */
   {
     #ifndef __mips__
+    #ifndef __aarch64__ // PTRACE_PEEKUSER returns -1 and set errno = 5 (EIO 5 Input/output error)
```
Comment (madscientist) [inline src/elfcore.c]: What are the ramifications of not setting this **user** value?

Also, did you see this note from the ptrace(2) man page:

>    C library/kernel differences
       At the system call level,  the  PTRACE_PEEKTEXT,  PTRACE_PEEKDATA,  and
       PTRACE_PEEKUSER requests have a different API: they store the result at
       the address specified by the data parameter, and the  return  value  is
       the  error  flag.  The glibc wrapper function provides the API given in
       DESCRIPTION above, with the result  being  returned  via  the  function
       return value.

So, maybe it didn't fail after all?

File: config.guess (hunk 1 above)
Comment (achichinov) [reply config.guess]: > I'm assuming that these changes were made when you used some different version of autoconf, and were not made by hand by you?

I copied the whole file, but it looks like I copied a modified version. And I don't recall where I copied it from. Most probably from github.
I just checked this file on GNU repo and it does not have these modifciations
https://git.savannah.gnu.org/gitweb/?p=config.git;a=blob_plain;f=config.guess;hb=45e181800a6a27268a9c5d79dcc60492fef9a9a0

I will update the file from the official repo.

File: src/elfcore.c [hunk 6]
```diff
@@ -1806,7 +1859,9 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   prpsinfo.pr_gid   = sys_getegid();
   prpsinfo.pr_pid   = main_pid;
   prpsinfo.pr_ppid  = sys_getppid();
+  #if !defined(__aarch64__)
```
Comment (madscientist) [inline src/elfcore.c]: Strange, why wouldn't getpgrp() not be available on ARM?  That's a bog-standard POSIX call that's been around forever.

File: src/elfcore.c [hunk 7]
```diff
@@ -1897,8 +1952,8 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   }
 
   /* scope */ {
-    int openmax  = sys_sysconf(_SC_OPEN_MAX);
-    int pagesize = sys_sysconf(_SC_PAGESIZE);
+    int openmax  = sysconf(_SC_OPEN_MAX);
+    int pagesize = sysconf(_SC_PAGESIZE);
```
Comment (madscientist) [inline src/elfcore.c]: Why this change?

File: src/linux_syscall_support_unittest.cc [hunk 8]
```diff
@@ -140,42 +140,73 @@ static void SigAction(int signum, siginfo_t *si, void *arg) {
...
+  #endif
   int signum       = SIGPWR;
   for (int info = 0; info < 2; info++) {
     signaled         = 0;
     struct kernel_sigaction sa = ZERO_SIGACT, old, orig;
+    #if defined(__aarch64__)
+    CHECK(!sys_rt_sigaction(signum, NULL, &orig, kSigsetSize));
+    #else
     CHECK(!sys_sigaction(signum, NULL, &orig));
+    #endif
```
Comment (madscientist) [inline src/linux_syscall_support_unittest.cc:154]: Is there a reason we can't just switch all the architectures to use rt_sigaction()?  It should work on Intel shouldn't it?

File: src/linuxthreads.c [hunk 9]
```diff
@@ -609,16 +618,22 @@ int ListAllProcessThreads(void *parameter,
...
 
     sys_sigprocmask(SIG_SETMASK, &sig_old, &sig_old);
 
     if (clone_pid >= 0) {
       int status, rc;
+      #if defined (__aarch64__)
+      while ((rc = sys_waitpid(clone_pid, &status, __WALL)) < 0 && ERRNO == EINTR) {
+             /* Keep waiting                                                 */
+      }
+      #else
```
Comment (madscientist) [inline src/linuxthreads.c]: Why does this need to be different?

File: src/linux_syscall_support.h [hunk 10]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (c) 2005-2008, Google Inc.
+/* Copyright (c) 2005-2011, Google Inc.
```
Comment (madscientist) [inline src/linux_syscall_support.h:1]: I didn't review this, I hope it's right :)

**Review by madscientist — APPROVED** [review]
Looks good although it seems like there are some simplifications that could be made to reduce the amount of ifdefs

File: src/elfcore.c (hunk 3 above)
Comment (achichinov) [reply src/elfcore.c]: > Strange. Why isn't this defined on Intel?

It is defined for Intel too in `<sys/procfs.h>` which was not included directly. But it was picked up on arm64 somehow.
I will include `<sys/procfs.h>` and remove this `#if`

File: src/elfcore.c (hunk 2 above)
Comment (achichinov) [reply src/elfcore.c]: I think this is good idea. I will do it.

File: src/elfcore.c (hunk 5 above)
Comment (achichinov) [reply src/elfcore.c]: I just realized that this is  `sys_ptrace` from `linux_syscall_support.h`. I tested the results on both platforms.
It returns data on x86 for some offsets and it always returns -1 (EIO) on arm64.
However, I will remove this `#ifdef` and revert the change back to the original code to avoid unforeseen results.

File: src/elfcore.c (hunk 7 above)
Comment (achichinov) [reply src/elfcore.c]: The latest `linux_syscall_support.h` does not provide `sys_sysconf()`

File: src/linux_syscall_support.h (hunk 10 above)
Comment (achichinov) [reply src/linux_syscall_support.h:1]: The version at 2011 is so perfect that no new versions are necessary. :-)

File: src/elfcore.c (hunk 6 above)
Comment (achichinov) [reply src/elfcore.c]: > Strange, why wouldn't getpgrp() not be available on ARM? That's a bog-standard POSIX call that's been around forever.

This is a syscall `sys_getpgrp()` from `LSS`. It is deprecated in the latest `asm-generic/unistd.h` but can be enabled via `__ARCH_WANT_SYSCALL_DEPRECATED`
I will enable it.

Comment (achichinov) [reply src/elfcore.c]: It turned out that `getpgrp()` always returns the PGID, so don't need to bother with the system calls from LSS and use `getpgrp()` on all platforms.

File: src/linuxthreads.c (hunk 9 above)
Comment (achichinov) [reply src/linuxthreads.c]: One is `sys_waitpid()` and another is `sys0_waitpid()`. I don't recall why I added It.
Original code works on the both platforms, so I will revert the change.

File: src/linux_syscall_support_unittest.cc (hunk 8 above)
Comment (achichinov) [reply src/linux_syscall_support_unittest.cc:154]: `linux_syscall_support_unittest` fails with SIGSEGV when calling `sys_rt_sigsuspend()` on Intel.
And it looks like `rt` functions can not be mixed with non `rt`.

File: src/elfcore.c [hunk 11]
```diff
@@ -923,6 +934,17 @@ static int CreateElfCore(void *handle,
                       }
                     }
 
+                    #if defined(__aarch64__)
+                    /* check if the flag is "pf" */
```
Comment (madscientist) [inline src/elfcore.c]: This comment is not helpful since it's obvious that this is what the code does.  Better would be to say WHY we check for this; apparently we don't dump segments with the "pf" flag for some reason?

File: src/elfcore.c [hunk 12]
```diff
@@ -1806,7 +1847,7 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   prpsinfo.pr_gid   = sys_getegid();
   prpsinfo.pr_pid   = main_pid;
   prpsinfo.pr_ppid  = sys_getppid();
-  prpsinfo.pr_pgrp  = sys_getpgrp();
+  prpsinfo.pr_pgrp  = getpgrp();
```
Comment (madscientist) [inline src/elfcore.c]: Just curious why we are using this rather than the sys_ version?  Maybe a comment here explaining will help me and future readers.

File: src/elfcore.c [hunk 13]
```diff
@@ -1897,8 +1938,8 @@ int InternalGetCoreDump(void *frame, int num_threads, pid_t *pids,
   }
 
   /* scope */ {
-    int openmax  = sys_sysconf(_SC_OPEN_MAX);
-    int pagesize = sys_sysconf(_SC_PAGESIZE);
+    int openmax  = sysconf(_SC_OPEN_MAX);
+    int pagesize = sysconf(_SC_PAGESIZE);
```
Comment (madscientist) [inline src/elfcore.c]: Ditto here: why do we remove sys?

File: src/linuxthreads.c [hunk 14]
```diff
@@ -316,9 +321,13 @@ static void ListerThread(struct ListerParams *args) {
     sa.sa_sigaction_ = SignalHandler;
     sys_sigfillset(&sa.sa_mask);
     sa.sa_flags      = SA_ONSTACK|SA_SIGINFO|SA_RESETHAND;
+    #if defined (__aarch64__)
+    sys_rt_sigaction(sync_signals[sig], &sa, (struct kernel_sigaction *)NULL, kSigsetSize);
+    #else
     sys_sigaction(sync_signals[sig], &sa, (struct kernel_sigaction *)NULL);
+    #endif
```
Comment (madscientist) [inline src/linuxthreads.c]: Is there a reason we can't use rt_sigaction() on intel (that is, just change this unconditionally to use rt_sigaction?)

it's OK if so, I just wonder why.

**Review by madscientist — APPROVED** [review]

**Review by madscientist — APPROVED** [review]
Looks good, thanks.  Are you ready for it to be pulled?

**Comment by achichinov** [comment]
> Looks good, thanks. Are you ready for it to be pulled?

Yes.
